### ISSUE TITLE
experiment with cffi API mode

### ISSUE THREAD

**Issue by jcupitt** [jcupitt/pyvips#22]
pyvips is very dynamic, and as a result, not very quick. 

We should experiment with cffi API mode, where a binary extension is built for the glue layer. It'd be much faster, but on the downside would need rebuilding occasionally.

**Comment by jcupitt** [comment]
Experimental branch here:

https://github.com/jcupitt/pyvips/tree/api

Seems to work, though it still needs some polish. It can work in ABI and API modes, but I'm not sure it'll fall back to ABI if API fails. It does start much faster. 

There are some problems with `g_free` as well: in API mode, you can't use lib functions directly as callback arguments, so it's forced to add a py layer to blob free. This then causes a segv on exit as libvips tries to call back into py after py has gone.

**Comment by jcupitt** [comment]
It knocks 20% off the test-suite run time, though it'll be less in regular code.

**Comment by jcupitt** [comment]
`g_free()` problems fixed with libvips 8.6.

It's not falling back correctly on install:

```
$ rm ~/vips/lib/pkgconfig/vips
vipsCC.pc    vips-cpp.pc  vips.pc      
$ rm ~/vips/lib/pkgconfig/vips* 
$ pkg-config vips --modversion
Package vips was not found in the pkg-config search path.
Perhaps you should add the directory containing `vips.pc'
to the PKG_CONFIG_PATH environment variable
No package 'vips' found
$ pip install --user -e .
Obtaining file:///home/john/GIT/pyvips
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/home/john/GIT/pyvips/setup.py", line 86, in <module>
```

**Comment by jcupitt** [comment]
Merged to master and released in 2.1
